### PR TITLE
docs/design: add proposal for plugin framework

### PR DESCRIPTION
**Pull request by lysu**
### What problem does this PR solve? 

Add proposal for plugin framework

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8802)

### CODE REVIEW

File: docs/design/2018-12-10-plugin-framework.md [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+- Last updated:  2018-12-10
+- Discussion at:
+
+## Abstract
+
+This proposal proposes to introduce the plugin framework to TiDB to support TiDB plugin development.
+
+## Background
+
+There are many cool customize requirements need be address but not convenient to merge to TiDB mainly repo, and Go 1.9+ introduce new plugin support, we can add plugin framework to TiDB to make them can be addressed, and attract more people to mutual build TiDB ecosystem.
```
Comment (jackysp) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
There are many cool customize requirements need be addressed but not convenient to merge to TiDB mainly repo, and Go 1.9+ introduce new plugin support, we can add plugin framework to TiDB to make them can be addressed, and attract more people to mutual build TiDB ecosystem.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+Go plugin doesn't support unload a plugin, but it can not stop us to load multiple version plugin into host program and framework ensure last reload one will be active, and others aren't unloaded but disable.
+
+So, we can reload plugin with a different version that is packaged by `pluginpkg` to modify plugin's implement logic, although we can not change the plugin's meta info(e.g. sysVars) now, I think it's still useful.
+
+#### Management
+
+For add a plugin to TiDB, we need:
+
+- add `-plugin-dir` as start argument to specify the folder contains plugins, e.g. '-plugin-dir=/data/deploy/tidb/plugin'
+- add `-plugin-load` as start argument to specify the plugin id(name "-" version) that need be load, e.g. '-plugin-load=conn_limit-1'
```
Comment (jackysp) [inline docs/design/2018-12-10-plugin-framework.md]: If no `-plugin-load` specified, does MySQL load all plugins in `-plugin-dir`?

File: docs/design/2018-12-10-plugin-framework.md [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+Add a plugin framework to TiDB.
+
+## Rationale
+
+Add the plugin framework base on Go's plugin support, but support uniform plugin manifest, package, and flexible SPI.
+
+plugin developer can build a TiDB plugin in 7 steps:
+
+- choose a plugin kind or no suitable need add new plugin kind
```
[the remark this answers is not in the document]
Comment (lysu) [reply docs/design/2018-12-10-plugin-framework.md]: @lilin90 I try to fix them via https://github.com/pingcap/tidb/pull/8802/commits/a9e14b18bd7f5ee4b49885b91eee8fda410aa764, maybe need take a look again thx~

File: docs/design/2018-12-10-plugin-framework.md (hunk 2 above)
Comment (lysu) [reply docs/design/2018-12-10-plugin-framework.md]: Yes, it doesn't. normally mysql default installed with plugin like version_token.so and so on but they are not load until `-plugin-load` in config or `INSTALL` command

**Comment by jackysp** [comment]
PTAL @tiancaiamao

File: docs/design/2018-12-10-plugin-framework.md [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+
+- Plugin name: we need to reload the plugin, so we need to load the same plugin with a different version that is at a much higher level than `pluginpath`.
+- Plugin version: plugin version makes it much easier to maintain.
+- Simple Dependence check: Go helps to use a check to build the version, but in the real world it is common that a plugin relies on b plugin's some new logic. We try to maintain a simple Required-Version relationship between different plugins.
+- Configuration: the plugin is a standalone module, so every plugin will introduce special sysVars just like normal MySQL variables. A user can use those variables to tweak plugin behaviors just like normal MySQL variables do.
+- Stats: the plugin will introduce new stats info. TiDB uses Prometheus, so the plugin can easily push metrics to Prometheus.
+- Plugin Category and Flexible SPI: TiDB can have limited plugin categories, the new plugin need choose one category it should be and implement SPI defined by those category.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
- Plugin Category and Flexible SPI: TiDB can have limited plugin categories, a new plugin need choose a category and implement SPI defined by those category.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+All of above construct the plugin metadata which we usually call `Manifest`. It describes the metadata and how others can use it by the plugin itself.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
All of above construct the plugin metadata which we usually call `Manifest`. `Manifest` describes the metadata and how others can use the plugin.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+We just use the Go plugin to load the TiDB plugin and the TiDB plugin gives us a `Manifest`. Then just use manifest to interact with the plugin. (Only load/lookup is heavy CGO call, and later call manifest is normal Golang method call.)
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
We just use the Go plugin mechanism to load the TiDB plugin and the TiDB plugin gives us a `Manifest`. Then just use manifest to interact with the plugin. (Only load/lookup is heavy CGO call, and later call manifest is normal Golang method call.)
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+
+The SPI (Service Provider Interface) for the plugin is the method that returns manifest and the manifest info itself. The method that returns manifest can be generated by `pluginpkg`, so implementing SPI work for the developer is to choose and construct different manifest info (`pluginpkg` also helps with this).
+
+`Manifest` is the base struct for all other sub manifests. The caller can use `Kind` and `DeclareXXManifest` to convert them to sub manifests.
+
+Manifest provides common metadata:
+
+- Kind: the plugin's category. Now we have the audit kind, authentication kind, and so on. It's easy to add more.
+- Name: name of the plugin, which is used to identify a plugin, so it cannot duplicate with other plugins.
+- Version: one plugin we can use to load multiple versions into TiDB, but just activate one of them to support hot-fix or hot-upgrade.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
- Version: we can load multiple versions a plugin into TiDB, but just activate one of them to support hot-fix or hot-upgrade.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+- RequireVersion: it will make a simple relationship between different plugins.
+- SysVars: it defines the plugin's configuration info.
+
+Manifest also provides three lifecycle extension points: 
+
+- Validate: called after loading all plugins but before onInit, so it can do cross plugins check before init.
+- OnInit: let plugin can prepare resource before real work.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
- OnInit: plugin can prepare resource before real work using OnInit.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+- OnShutDown: let plugin can clean up some outer resources before dying.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
- OnShutDown: plugin can clean up its resources before dying using OnShutDown.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+    SysVars        map[string]*variable.SysVar
+    Validate       func(ctx context.Context, manifest *Manifest) error
+    OnInit         func(ctx context.Context) error
+    OnShutdown     func(ctx context.Context) error
+}
+```
+
+Base on `Kind`, we can define other subManifest for authentication plugin, audit plugin and so on.
+
+Every subManifest will have a `Manifest` anonymous field as the FIRST field in struct definition, so every subManifest can be used as `Manifest` (by `unsafe.Pointer` cast). For example, an audit plugin' manifest will be like this:
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]: `unsafe.Pointer` is really tricky

File: docs/design/2018-12-10-plugin-framework.md [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+- `version`: the version of a plugin. For the same plugin, the same version is only loaded once.
+- `description`: description of plugin usage.
+- `license`: license of the plugin, which will display in `show plugins`.
+- `sysVars`: it defines the variable needed by this plugin with name, scope and default value.
+- `validate`: it specifies the callback function used to validate before load, e.g. auth plugin check `-with-skip-grant-tables` configuration.
+- `onInit`: it specifies the callback function used to init plugin before it joins real work.
+- `onShutdown`: the callback function will be called when the plugin shuts down to release outer resource held by the plugin, normally TiDB shutdown.
+- `export`: it defines the callback list for the special kind plugins, e.g. for auth plugin it uses a `NotifyEvent` method to implement the `notifyEvent` extension point.
+
+`pluginpkg` will generate code and build as the Go plugin, and using plugin packagd we also control the plugin binary's format:
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
`pluginpkg` generates code and the generated code is built as a Go plugin, and using plugin package we also control the plugin binary's format:
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+
+- The plugin file name is `[pluginName]-[version].so`, so we can know the plugin's version from the filename.
+- `pluginpath` will be `[pluginName]-[version]`, and then we can load the same plugin in a different version in the same host program.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]:
```suggestion
- `pluginpath` will be `[pluginName]-[version]`, and then we can load the same plugin of a different version in the same host program.
```

File: docs/design/2018-12-10-plugin-framework.md [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+
+Plugin's variable name must use plugin name as the prefix. At last, the plugin cannot be reloaded if we change the plugin's sysVar (include default-value, add or remove variable).
+
+We implement it by adding the plugin variable into `variable.SysVars` before `bootstrap`, so later `doDMLWorker` will handle them just as normal sysVars, and change `loadCommonGlobalVarsSQL` to load them. (that it cannot unload plugin and cannot modify sysVar during reload makes this implementation easier) 
+
+#### Dependency
+
+Go's plugin mechanism will check all dependency package hash to ensure link time and run time use the same version([see code](https://github.com/golang/go/blob/50bd1c4d4eb4fac8ddeb5f063c099daccfb71b26/src/runtime/plugin.go#L52)), so we no longer need to care about compiling package dependency.
+
+But for the real world, there may be logic dependency between plugins. For example, some guy writes an authorization plugin but it relies on vault plugin and only works when vault is enabled but does not directly rely on the vault plugin's source code.
```
Comment (tiancaiamao) [inline docs/design/2018-12-10-plugin-framework.md]: vault?

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**